### PR TITLE
chipsec: init at 1.3.7

### PR DESCRIPTION
**Pull request by JohnAZoidberg**
###### Motivation for this change
The Nix(OS) community should check their firmware and its settings for vulnerabilities!

It's included twice: The top-level attribute doesn't include the kernel driver and can be executed on any OS with Python, and the one in linuxPackages which includes the Linux kernel driver and is therefore able to do more advanced checks.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by joachifm** [comment]
@GrahamcOfBorg build chipsec linuxPackages.chipsec

**Comment by JohnAZoidberg** [comment]
Oops, ofBorg failed to build it because it doesn't work on ARM.

File: pkgs/tools/security/chipsec/default.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      It includes a security test suite, tools for accessing various low level
+      interfaces, and forensic capabilities. It can be run on Windows, Linux,
+      Mac OS X and UEFI shell.
+    '';
+    license = licenses.gpl2;
+    homepage = https://github.com/chipsec/chipsec;
+    maintainers = with maintainers; [ johnazoidberg ];
+    # This package description is currently only able to build the Linux driver.
+    # But the other functionality should work on all x86 platforms.
+    platforms = [ "x86_64-linux" ] ++ lib.optional (isNull kernel) "x86_64-darwin";
```
Comment (joachifm) [inline pkgs/tools/security/chipsec/default.nix]: Nit: to me `if withDriver then [ "x86_64-linux" ] else platforms.all` is clearer & makes the comment entirely redundant.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -684,6 +684,9 @@ in
 
   chezmoi = callPackage ../tools/misc/chezmoi { };
 
+  # Without kernel driver, should build and work on non-linux as well
```
Comment (joachifm) [inline pkgs/top-level/all-packages.nix]: This comment somewhat disagrees with `meta.platforms` which is restricted to x64-linux & macos (which in turn is inconsistent with the package description which also mentions windows). I think its better to remove this comment & just let `meta.platforms` speak by itself.

**Comment by joachifm** [comment]
LGTM. Added a few nits for your consideration.

**Comment by joachifm** [comment]
Another thing that occurred to me is, if I add both variants to my system will I end up with two copies of the userland tools in the system closure? It seems to me like I will, as different flags would produce different store paths. Perhaps the `withDriver` case ought to build *only* the driver. On the other hand, I'd hope hardlinking would resolve the space inefficiency ... so maybe not worth the effort.

**Comment by JohnAZoidberg** [comment]
Hmm, but it loads the driver on demand. We'd have to make sure it can find it if it's in a different derivation. I don't think the package is designed to be able to look for the driver elsewhere, I can look into it.
I didn't know a better way to make the driver optional other than including it in linuxPackages and top-level. Is there?

**Comment by joachifm** [comment]
By loading on demand, I expect you mean calls modprobe? If so it should work as long as the package is added to `boot.extraModulePackages` (and the call to modprobe/insmod does not hardcode the module path). If it does hardcode the path, that can be patched out.

I think I was wrong to raise the concern about duplicated stuff in the system closure, I have not looked but `extraModulePackages` probably only copies kernel modules anyway. There will still be some wasted bandwidth & work in the linux driver+userspace scenario, however. This could be solved by simply defaulting `withDriver = true` when building for linux.

**Comment by JohnAZoidberg** [comment]
Yes exactly, calling modprobe.
But if we rely on `boot.extraModulePackages` we'd have to make it a module which is not the right thing to do because using chipsec is a one-off kind of thing.

But can you just default to building the driver on Linux? I don't think you can, because you also need to get the exact kernel passed that used on the system. Hence `linuxPackagesFor` to build it for a specific kernel version.

Is there precedent? Is there another package that's not just a kernel module but a program that brings its own kernel module that we package but don't have a NixOS module for? P.S. Looks like `sysdig` is in a similar situation.

Otherwise I think it's fine like this. We'll have to hope that people are looking for the `linuxPackages.chipsec` attribute if they notice that `chipsec` doesn't include the driver.

File: pkgs/tools/security/chipsec/default.nix (hunk 1 above)
Comment (JohnAZoidberg) [reply pkgs/tools/security/chipsec/default.nix]: You're right. That does make the most sense.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (JohnAZoidberg) [reply pkgs/top-level/all-packages.nix]: I made it clearer and explicitly pass no kernel here.

**Comment by JohnAZoidberg** [comment]
Thanks for discussing the details!

**Pull request merged**